doc/tests-updates.md

```diff
@@ -0,0 +1,381 @@
+the request.
+
+- Cache tests. It's important whether a server connection was used to serve
+the request or not.
+
+- Request-response chain integrity. Need to validate that every client receives
+its responses in the correct order.
+
+- Fault injections. Based on other types of tests, but random errors to be
+emulated using SystemTap or eBPF.
+
+- Fussing. Generating random requests and responses, Tempesta must stay
+operational and serve as many requests as possible.
+
+- Workload testing. Generate pretty high request rate to confirm data safety
+on highly concurrent code execution. The more CPUs is used for Tempesta -
+the better. But CPU utilisation must be less than 80% to assure that no extreme
+conditions happen.
+
+- Stress testing. Testing under extreme conditions, when request rate is higher
+than Tempasta can serve.
+
+- Low level TCP testing. Some features can affect TCP behaviour, so need to
+validate that the Tempesta conforms RFCs and doesn't break kernel behaviour.
+
+- TLS testing. Having in-house TLS implementation gives great powers and great
+responsibility. We must assure TLS support messages integrity, encryption
+integrity. It differs from other tests, since it's important to work with raw
+data lying under HTTP on client and server side.
+
+- Tempesta-originated requests. Pretty specific tests, since Tempesta may
+generate requests on its own, e.g. background cache revalidation, or health
+monitor requests. It differs from other tests, since it's important to validate
+support TCP messages.
+
+
+There also some requirements on tests process:
+
+- Full, reduced and extra-stability set of tests. Reduced for developers to
+run before making pull requests. Full to run against pull requests.
+Extra-stability for every night testing to assure stability under unpredictable
+loads.
+
+- Easy new test addition.
+
+- Easy test case failure analysis.
+
+- Generate a few tests based on singe test description. E.g. in fussing test
+it's required to know, how the Tempesta would react on a message, and what
+should be threated as error. The 'splitting on different skbs' is actually
+fussing test feature, not functional. Usually it also make sense to generate
+workload test based on functional profile.
+
+
+## Problems of current implementation
+
+The main problem is having two frameworks in the same time. Old, feature rich,
+but too sophisticated, with requirement to understand 5-6 files to understand
+what the test is actually does. And a new, more friendly, but not complete and
+with lack of the most important features. Transition of an old one to a new
+wasn't completed, and the development of both was rather unorganized, thus
+new framework is becoming more and more complex and loses its advantages.
+
+In both frameworks every test type is implemented as unique test, so it's
+required to write message integrity, fussing, workload and many other test
+cases for the same situation.
```

Comment:
I still have no idea how to easily make the test framework to split a message into different sets of skbs. Probably we can use `TCP_NODELAY` and write the message in different parts, but I'm not sure about bufferization on Python side. However, I believe the task is solvable in generic way.

I suppose the the message integrity and probably skb splitting can be done on deproxy side. If we need the same for Nginx tests, then we can just implement the logic as a separate class/module.

All in all, with https://www.youtube.com/watch?v=oO-FMAdjY68 and the requirements above in mind, I believe that the test framework should use OO relatively complex class inheritance for the helping functionality while the tests should just on/off necessary flags for message integrity calculation, sending in different skb and other things. I.e. the helping framework must provide as declarative API as possible and it's not so important how complex it internally is.